vopr: implement pauses

matklad opened the pull request:
Adds a new failure mode for VOPR --- pausing a machine. This can uncover
new interesting interleaving and also models VM migration scenario.

To model the latter case faithfully, we tick the time even if the
machine is paused.

This is vopr-lite green a 450_357 seeds

matklad committed:
vopr: restructure the code to allow for pausing replicas

- add .paused state to replica health, which is a substate of up
- restructure the tick loop to make it easy to tick only unpause replicas

matklad committed:
vopr: implement pauses

Adds a new failure mode for VOPR --- pausing a machine. This can uncover
new interesting interleaving and also models VM migration scenario.

To model the latter case faithfully, we tick the time even if the
machine is paused.

matklad committed:
vopr: big-endian naming

sentientwaffle commented:
CFO found a crash:

```
./zig/zig build -Drelease vopr -- --lite 657199375419925836
```

I'm actually surprised that we haven't seen this bug before! It looks like we are recovering from a crash before any op is prepared, and our root op is corrupt, so we don't have a proper op-head.

<details>
<summary>Logs</summary>

```
[warn] (journal): 1: recover_slot: recovered slot=0000 label=@C decision=vsr operation=vsr.Operation.reserved op=0 view=0
[debug] (journal): 1: recover_slot: recovered slot=0001 label=@I decision=nil operation=vsr.Operation.reserved op=1 view=0
[debug] (journal): 1: recover_slot: recovered slot=0002 label=@I decision=nil operation=vsr.Operation.reserved op=2 view=0
[debug] (journal): 1: recover_slot: recovered slot=0003 label=@I decision=nil operation=vsr.Operation.reserved op=3 view=0
[debug] (journal): 1: recover_slot: recovered slot=0004 label=@I decision=nil operation=vsr.Operation.reserved op=4 view=0
[debug] (journal): 1: recover_slot: recovered slot=0005 label=@I decision=nil operation=vsr.Operation.reserved op=5 view=0
[debug] (journal): 1: recover_slot: recovered slot=0006 label=@I decision=nil operation=vsr.Operation.reserved op=6 view=0
[debug] (journal): 1: recover_slot: recovered slot=0007 label=@I decision=nil operation=vsr.Operation.reserved op=7 view=0
[debug] (journal): 1: recover_slot: recovered slot=0008 label=@I decision=nil operation=vsr.Operation.reserved op=8 view=0
[debug] (journal): 1: recover_slot: recovered slot=0009 label=@I decision=nil operation=vsr.Operation.reserved op=9 view=0
[debug] (journal): 1: recover_slot: recovered slot=0010 label=@I decision=nil operation=vsr.Operation.reserved op=10 view=0
[warn] (journal): 1: recover_slot: recovered slot=0011 label=@B decision=vsr operation=vsr.Operation.reserved op=11 view=0
[debug] (journal): 1: recover_slot: recovered slot=0012 label=@I decision=nil operation=vsr.Operation.reserved op=12 view=0
[debug] (journal): 1: recover_slot: recovered slot=0013 label=@I decision=nil operation=vsr.Operation.reserved op=13 view=0
[warn] (journal): 1: recover_slot: recovered slot=0014 label=@B decision=vsr operation=vsr.Operation.reserved op=14 view=0
[debug] (journal): 1: recover_slot: recovered slot=0015 label=@I decision=nil operation=vsr.Operation.reserved op=15 view=0
[debug] (journal): 1: recover_slot: recovered slot=0016 label=@I decision=nil operation=vsr.Operation.reserved op=16 view=0
[debug] (journal): 1: recover_slot: recovered slot=0017 label=@I decision=nil operation=vsr.Operation.reserved op=17 view=0
[debug] (journal): 1: recover_slot: recovered slot=0018 label=@I decision=nil operation=vsr.Operation.reserved op=18 view=0
[debug] (journal): 1: recover_slot: recovered slot=0019 label=@I decision=nil operation=vsr.Operation.reserved op=19 view=0
[debug] (journal): 1: recover_slot: recovered slot=0020 label=@I decision=nil operation=vsr.Operation.reserved op=20 view=0
[debug] (journal): 1: recover_slot: recovered slot=0021 label=@I decision=nil operation=vsr.Operation.reserved op=21 view=0
[debug] (journal): 1: recover_slot: recovered slot=0022 label=@I decision=nil operation=vsr.Operation.reserved op=22 view=0
[debug] (journal): 1: recover_slot: recovered slot=0023 label=@I decision=nil operation=vsr.Operation.reserved op=23 view=0
[debug] (journal): 1: recover_slot: recovered slot=0024 label=@I decision=nil operation=vsr.Operation.reserved op=24 view=0
[debug] (journal): 1: recover_slot: recovered slot=0025 label=@I decision=nil operation=vsr.Operation.reserved op=25 view=0
[debug] (journal): 1: recover_slot: recovered slot=0026 label=@I decision=nil operation=vsr.Operation.reserved op=26 view=0
[debug] (journal): 1: recover_slot: recovered slot=0027 label=@I decision=nil operation=vsr.Operation.reserved op=27 view=0
[debug] (journal): 1: recover_slot: recovered slot=0028 label=@I decision=nil operation=vsr.Operation.reserved op=28 view=0
[debug] (journal): 1: recover_slot: recovered slot=0029 label=@I decision=nil operation=vsr.Operation.reserved op=29 view=0
[debug] (journal): 1: recover_slot: recovered slot=0030 label=@I decision=nil operation=vsr.Operation.reserved op=30 view=0
[debug] (journal): 1: recover_slot: recovered slot=0031 label=@I decision=nil operation=vsr.Operation.reserved op=31 view=0
[debug] (journal): 1: recover_slots: dirty=3 faulty=3
thread 23949 panic: attempt to use null value
/var/home/djg/C/t/db/docs/src/vsr/replica.zig:740:69: 0x11e2b3b in open (vopr)
            const header_head = self.journal.header_with_op(self.op).?;
                                                                    ^
/var/home/djg/C/t/db/docs/src/testing/cluster.zig:566:29: 0x11e6f41 in replica_open (vopr)
            try replica.open(
                            ^
/var/home/djg/C/t/db/docs/src/testing/cluster.zig:506:37: 0x1293884 in replica_restart (vopr)
            try cluster.replica_open(replica_index, .{
                                    ^
/var/home/djg/C/t/db/docs/src/vopr.zig:1158:42: 0x1236212 in replica_restart (vopr)
        simulator.cluster.replica_restart(
                                         ^
/var/home/djg/C/t/db/docs/src/vopr.zig:1093:34: 0x1292cc2 in tick_crash_down (vopr)
        simulator.replica_restart(replica.replica, fault);
                                 ^
/var/home/djg/C/t/db/docs/src/vopr.zig:1037:51: 0x123536d in tick_crash (vopr)
                .down => simulator.tick_crash_down(replica),
                                                  ^
/var/home/djg/C/t/db/docs/src/vopr.zig:596:29: 0x11fe7bb in tick (vopr)
        simulator.tick_crash();
                            ^
/var/home/djg/C/t/db/docs/src/vopr.zig:324:23: 0x11fb939 in main (vopr)
        simulator.tick();
                      ^
/var/home/djg/C/bin/zig-linux-x86_64-0.13.0/lib/std/start.zig:524:37: 0x11a7b75 in posixCallMainAndExit (vopr)
            const result = root.main() catch |err| {
                                    ^
/var/home/djg/C/bin/zig-linux-x86_64-0.13.0/lib/std/start.zig:266:5: 0x11a7691 in _start (vopr)
    asm volatile (switch (native_arch) {
    ^
```
</details>

sentientwaffle commented on src/vsr/replica_test.zig:
```diff
@@ -2095,10 +2095,15 @@ const TestReplicas = struct {
-    pub fn health(t: *const TestReplicas) ReplicaHealth {
-        var value_all: ?ReplicaHealth = null;
+    const Health = enum { up, down };
+
+    pub fn health(t: *const TestReplicas) Health {
+        var value_all: ?Health = null;
         for (t.replicas.const_slice()) |r| {
-            const value = t.cluster.replica_health[r];
+            const value: Health = switch (t.cluster.replica_health[r]) {
+                .up => .up,
+                .down => .down,
+            };
```
Alternatively, set `Health` to `std.meta.Tag(ReplicaHealth)`, then it can just cast the union into an enum instead of `switch`ing.

matklad replied:
I considered that, and decided that it's better to have ` enum { up, down };` spelled here explicitly:

* this way, the reader won't have to goto definition to ReplicaHealth to understand what this thing is about
* philosophically, I think it's good if actual tests don't touch the underlying APIs directly, and instead interact strictly with TestContext's own API, to minimize the need to refactor the tests when refactoring non-test code. 

But I am fine either way!

matklad committed:
vsr: fix a crash when the root prepare gets corrupted

On this seed, a replica is on the initial checkpoint, and everything it
prepared (3 prepares in total) got corrupted, including the root
prepare.

Normally in this case the replica uses op_checkpoint for op_head, but,
as `self.journal.op_maximum() < self.op_checkpoint()` is `0 < 0` is
false, we skip the relevant code block, and end up in a situation when
`op_head == 0 and journal.header_with_op(0) == null`, tripping an
assert.

The root cause is that `op_maximum` should really be returning `null`
here, rather than `0`, to disambiguate the case where `0` is indeed the
maximum op, and the case where we don't have a maximum op at all.

I considered changing its return type, but then I'll also have to change
the return type of `op_maximum_untrusted`, and add a bunch of special
cases for no real benefit.

So, instead I just special-case a single if.

I don't understand why this was uncovered by pausing (parent commit) ---
I _think_ this should be findable on main as well, but perhaps with
lower probability.

Seed: ./zig/zig build vopr -Drelease -- --lite 657199375419925836

matklad commented:
Fixed the vopr failure, which feels like it should be unrelated to pauses. Sorry for taking too long to turn this around!

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -715,7 +715,10 @@ pub fn ReplicaType(
-                    if (self.journal.op_maximum() < self.op_checkpoint()) {
+                    if (self.journal.op_maximum() < self.op_checkpoint() or
+                        // Corrupted root prepare:
+                        (self.journal.op_maximum() == 0 and self.journal.header_with_op(0) == null))
```
We _could_ farther restrict this clause to `op_checkpoint() == 0`, though that is sort of implied by `op_maximum == 0`, and since we didn't exit earlier due to `self.journal.faulty.count == constants.journal_slot_count`.

matklad replied:
> though that is sort of implied by op_maximum == 0, and since we didn't exit earlier due to self.journal.faulty.count == constants.journal_slot_count.

This follows more directly --- if we get to the second arm of this `or`, we know that `op_maximum >= op_checkpoint`, so checkpoint can only be zero. 

Actually, I think the `faulty=slot_count` implication would be wrong --- our journal is larger than a checkpoint, so, it could be the case that:

* you are one the _first_ checkpoint
* all _real_ prepares are corrupted
* but there are some valid reserved slots still

In this case, our op_maximum would be 0, and we'll enter this branch via `self.journal.op_maximum() < self.op_checkpoint()` condition
